metadata/src/lib.rs

```diff
@@ -0,0 +1,458 @@
+                visited_ids,
+            ));
+            bytes
+        }
+        TypeDef::Tuple(tuple) => {
+            let mut bytes = Vec::new();
+            for field in tuple.fields() {
+                bytes.extend(get_type_hash(registry, field.id(), visited_ids));
+            }
+            bytes
+        }
+        TypeDef::Primitive(primitive) => {
+            let mut bytes = Vec::new();
+            primitive.encode_to(&mut bytes);
+            bytes
+        }
+        TypeDef::Compact(compact) => {
+            let mut bytes = Vec::new();
+            bytes.extend(get_type_hash(
+                registry,
+                compact.type_param().id(),
+                visited_ids,
+            ));
+            bytes
+        }
+        TypeDef::BitSequence(bitseq) => {
+            let mut bytes = Vec::new();
+            bytes.extend(get_type_hash(
+                registry,
+                bitseq.bit_order_type().id(),
+                visited_ids,
+            ));
+            bytes.extend(get_type_hash(
+                registry,
+                bitseq.bit_store_type().id(),
+                visited_ids,
+            ));
+            bytes
+        }
+    };
+    bytes.extend(data);
+    hash(&bytes)
+}
+
+/// Obtain the hash representation of a `scale_info::Type` identified by id.
+fn get_type_hash(
+    registry: &PortableRegistry,
+    id: u32,
+    visited_ids: &mut HashSet<u32>,
+) -> [u8; 32] {
+    let ty = registry.resolve(id).unwrap();
+
+    let mut bytes = vec![MetadataHashableIDs::Type as u8];
+    ty.path().segments().encode_to(&mut bytes);
+    // Guard against recursive types
+    if !visited_ids.insert(id) {
+        return hash(&bytes)
+    }
+
+    let ty_def = ty.type_def();
+    bytes.extend(get_type_def_hash(registry, ty_def, visited_ids));
+
+    hash(&bytes)
+}
+
+/// Obtain the hash representation of a `frame_metadata::ExtrinsicMetadata`.
+fn get_extrinsic_hash(
+    registry: &PortableRegistry,
+    extrinsic: &ExtrinsicMetadata<PortableForm>,
+) -> [u8; 32] {
+    let mut visited_ids = HashSet::<u32>::new();
+    let mut bytes = vec![MetadataHashableIDs::Extrinsic as u8];
+
+    bytes.extend(get_type_hash(registry, extrinsic.ty.id(), &mut visited_ids));
+    bytes.push(extrinsic.version);
+    for signed_extension in extrinsic.signed_extensions.iter() {
+        signed_extension.identifier.encode_to(&mut bytes);
+        bytes.extend(get_type_hash(
+            registry,
+            signed_extension.ty.id(),
+            &mut visited_ids,
+        ));
+        bytes.extend(get_type_hash(
+            registry,
+            signed_extension.additional_signed.id(),
+            &mut visited_ids,
+        ));
+    }
+
+    hash(&bytes)
+}
+
+/// Obtain the hash representation of a `frame_metadata::PalletMetadata`.
+pub fn get_pallet_hash(
+    registry: &PortableRegistry,
+    pallet: &frame_metadata::PalletMetadata<PortableForm>,
+    cache: &mut MetadataHasherCache,
```

Comment:
Since I think we said that the cache has to be per-pallet now, is there still a reason to pass it in, or would it be better to create it at the beginning of this function?

Reply:
Aah ok, so looking at this function, all it actually does with the cache is check whether the pallet we're asking to generate a hash for already has one. I'd suggest that that check could live entirely outside of this metadata crate; leave it up to the caller to decide whether or not to call this function, and don't pass in this cache to it :)

Something we might want (but it can be a follow-up PR once we benchmark how long this hashing takes) is an internal cache like `HashMap<u32, Hash>` so that we don't need to hash a given type more than once. If this cache is created at the beginning of this `get_pallet_hash` call and passed in to things, it'll be a per-pallet cache only, which means you won't hit those issues you had with recursive type IDs I think :)

Reply:
This might help to cache sequential calls to the pallet, having an early return. Another benefit for this is the scenario in which we do a full metadata check, which will hash each pallet 😄 
Although I think we can do the cache entirely on the Client / inside the RuntimeAPI, we might still do double work for that scenario.

I think having the cache entirely in the user's hand would be a good trade-off here and removing the cache from `get_pallet_hash` and metadata check. It all depends on how users are going to do the checks. I expect the user to check with a high frequency either the full metadata or a small subset of pallets, as opposed to a combination of both 😄